Feature | `graph init` Add support to download spkg files from spkg.io

joshuanazareth97 opened the pull request:
## Change Description

- Allow passing a file url for spkg files instead of just a file path.
- Download the file if a url has been passed

## Related Issue / Discussion
https://github.com/graphprotocol/graph-tooling/issues/1641

## Screenshot of change
![image](https://github.com/graphprotocol/graph-tooling/assets/31418899/988fcabe-c7f3-42d4-86bc-c9b6f3e11f32)
![image](https://github.com/graphprotocol/graph-tooling/assets/31418899/3980d49e-395f-4069-b987-0e942b243450)
![image](https://github.com/graphprotocol/graph-tooling/assets/31418899/71a6cbe9-4aba-44b4-b27c-163fbdf27f19)

joshuanazareth97 commented:
Hi @saihaj can you please review this PR? Let me know if there's anything you need me to do before you merge it. Thanks!

saihaj commented:
thanks @joshuanazareth97 sorry for the delay. I was away on vacation last week, catching up on things will try to land this early next week

saihaj commented on .changeset/early-walls-talk.md:
```diff
@@ -0,0 +1,5 @@
+---
+'@graphprotocol/graph-cli': patch
+---
+
+Order list of evm chains in graph init command
```
```suggestion
---
'@graphprotocol/graph-cli': minor
---

`graph init` to allow downloading `spkg.io` files 
```

saihaj commented on packages/cli/src/command-helpers/download.ts:
```diff
@@ -0,0 +1,13 @@
+import { createWriteStream } from 'fs';
+import { http } from 'gluegun';
+
+export async function downloadFile(fileUrl: string, outputLocationPath: string) {
+  const writer = createWriteStream(outputLocationPath);
+  const api = http.create({
+    baseURL: fileUrl,
+  });
+  return api.get('', {}, { responseType: 'stream' }).then((response: any) => {
+    response.data.pipe(writer);
+    return Promise.resolve(outputLocationPath);
+  });
+}
```
I would use `fetch`, we maintain an internal wrapper that also passes some headers that can let origin servers know where the traffic is coming from.

https://github.com/graphprotocol/graph-tooling/blob/1f8d86f10ceee6f8e34ba942cfa76a37f1f44a42/packages/cli/src/fetch.ts#L5

joshuanazareth97 replied:
Have now resolved this to use the wrapped fetch

saihaj commented on packages/cli/src/command-helpers/spkg.ts:
```diff
@@ -0,0 +1,18 @@
+import { filesystem } from 'gluegun';
+
+export const isSpkgUrl = (value: string) => {
+  return value.startsWith('https://spkg.io/streamingfast');
```
don't think we need `/streamingfast` on the subpath (for example we have semiotic on this path https://substreams.dev/semiotic-ai/uniswap-x/v0.1.0#modules)  we just need to make sure it is from `spkg.io`

we want to allow users either passing `https://spkg.io/semiotic-ai/uniswap-x-v0.1.0.spkg` or `spkg.io/semiotic-ai/uniswap-x-v0.1.0.spkg` and it should work



```suggestion
  return value.startsWith('spkg.io');
```

joshuanazareth97 replied:
Updating the check to 
```
value.startsWith(`https://spkg.io`) || value.startsWith(`spkg.io`)
```

saihaj reviewed (changes requested):
few minor nits, thanks for adding tests!

(sorry for the delayed review)

saihaj commented:
@joshuanazareth97 are you still working on this?

joshuanazareth97 commented:
> @joshuanazareth97 are you still working on this?

@saihaj apologies, was occupied with some other stuff last week. Have addressed all your comments.

joshuanazareth97 commented:
@saihaj is there anything else you need me to do here? Can we merge this PR if the changes are made to your satisfaction?

YaroShkvorets commented:
Implemented in #1788
